src/settings/factory_reset.c

```diff
@@ -0,0 +1,305 @@
+                                   wallet_list_t *wallets_in_card2,
+                                   wallet_list_t *common_wallets) {
+  uint8_t common_wallet_count = 0;
+  memzero(common_wallets, sizeof(wallet_list_t));
+
+  for (uint8_t index = 0; index < wallets_in_card1->count; index++) {
+    bool match = false;
+
+    for (uint8_t i = 0; i < wallets_in_card2->count; i++) {
+      if (memcmp(wallets_in_card1->wallet[index].name,
+                 wallets_in_card2->wallet[i].name,
+                 NAME_SIZE) == 0 &&
+          memcmp(wallets_in_card1->wallet[index].id,
+                 wallets_in_card2->wallet[i].id,
+                 WALLET_ID_SIZE) == 0) {
+        common_wallet_count += 1;
+        match = true;
+        break;
+      }
+    }
+
+    // If the metadata matched, store it in the common list
+    if (match) {
+      memcpy(&common_wallets->wallet[common_wallet_count - 1],
+             &wallets_in_card1->wallet[index],
+             sizeof(wallet_metadata_t));
+    }
+  }
+
+  common_wallets->count = common_wallet_count;
+  return;
+}
+
+static bool get_wallet_list_from_two_cards(wallet_list_t *wallet_list) {
+  wallet_list_t wallets_in_card1 = {0};
+  wallet_list_t wallets_in_card2 = {0};
+
+  card_fetch_wallet_list_config_t configuration = {0};
+  card_fetch_wallet_list_response_t response = {0};
+
+  configuration.acceptable_cards = 0xF;
+  configuration.heading = ui_text_tap_1_2_cards;
+  configuration.msg = ui_text_place_card_below;
+  configuration.skip_card_removal = false;
+
+  response.wallet_list = &wallets_in_card1;
+  response.tapped_card = 0;
+
+  if (!card_fetch_wallet_list(&configuration, &response)) {
+    return false;
+  }
+
+  // Do not accept the same card again
+  configuration.acceptable_cards ^= response.tapped_card;
+  configuration.heading = ui_text_tap_2_2_cards;
+  configuration.skip_card_removal = true;
+
+  response.wallet_list = &wallets_in_card2;
+  response.tapped_card = 0;
+
+  if (!card_fetch_wallet_list(&configuration, &response)) {
+    return false;
+  }
+
+  get_common_wallet_list(&wallets_in_card1, &wallets_in_card2, wallet_list);
+  return true;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void factory_reset(void) {
+  wallet_list_t wallets_in_cards = {0};
+  wallet_list_t wallets_in_vault = {0};
+
+  uint8_t valid_wallets = get_valid_wallet_meta_data_list(&wallets_in_vault);
+
+  if (0 < valid_wallets &&
+      !core_scroll_page(NULL, ui_text_factory_reset_instruction, NULL)) {
+    return;
+  }
+
+  if (!core_scroll_page(NULL, ui_text_confirm_factory_reset, NULL)) {
+    return;
+  }
+
+  // If there is atleast one valid wallet on the device, ensure that it is
+  // present in atleast 2 X1 cards
+  if (0 < valid_wallets) {
+    if (!get_wallet_list_from_two_cards(&wallets_in_cards)) {
+      return;
```

Comment:
let's ensure screen flicker/stuck behavior does not appear

Reply:
Added